abe/cpabe/tkn20/internal/tkn/tk.go

```diff
@@ -0,0 +1,809 @@
+	if err != nil {
+		return nil, fmt.Errorf("c1 serializing: %w", err)
+	}
+	ret = appendLenPrefixed(ret, c1Bytes)
+
+	// Now we need to indicate how long c2, c3, c3neg are.
+	// Each array will be the same size (or nil), so with more work we can specalize
+	// but for now we will ignore that.
+
+	c2Len := len(hdr.c2)
+
+	ret = append(ret, 0, 0)
+	binary.LittleEndian.PutUint16(ret[len(ret)-2:], uint16(c2Len))
+	for i := 0; i < c2Len; i++ {
+		c2dat, errM := hdr.c2[i].marshalBinary()
+		if errM != nil {
+			return nil, fmt.Errorf("c2 serializing %d: %w", i, errM)
+		}
+		ret = appendLenPrefixed(ret, c2dat)
+	}
+	c3Len := len(hdr.c3)
+	ret = append(ret, 0, 0)
+	binary.LittleEndian.PutUint16(ret[len(ret)-2:], uint16(c3Len))
+	for i := 0; i < c3Len; i++ {
+		c3dat, errM := hdr.c3[i].marshalBinary()
+		if errM != nil {
+			return nil, fmt.Errorf("c3 serializing %d: %w", i, errM)
+		}
+		ret = appendLenPrefixed(ret, c3dat)
+	}
+	for i := 0; i < c3Len; i++ {
+		var c3negdat []byte
+		if hdr.c3neg[i] != nil {
+			c3negdat, err = hdr.c3neg[i].marshalBinary()
+			if err != nil {
+				return nil, fmt.Errorf("c3neg serializing %d: %w", i, err)
+			}
+		} else {
+			c3negdat = nil
+		}
+		ret = appendLenPrefixed(ret, c3negdat)
+	}
+	return ret, nil
+}
+
+func (hdr *ciphertextHeader) unmarshalBinary(data []byte) error {
+	pBytes, data, err := removeLenPrefixed(data)
+	if err != nil {
+		return err
+	}
+	hdr.p = new(Policy)
+	err = hdr.p.UnmarshalBinary(pBytes)
+	if err != nil {
+		return err
+	}
+	c1Bytes, data, err := removeLenPrefixed(data)
+	if err != nil {
+		return err
+	}
+	hdr.c1 = newMatrixG2(0, 0)
+	err = hdr.c1.unmarshalBinary(c1Bytes)
+	if err != nil {
+		return err
+	}
+	c2Len := int(binary.LittleEndian.Uint16(data))
+	hdr.c2 = make([]*matrixG2, c2Len)
+	data = data[2:]
+	var c2data []byte
+	var c3data []byte
+	var c3negdata []byte
+
+	for i := 0; i < c2Len; i++ {
+		c2data, data, err = removeLenPrefixed(data)
+		if err != nil {
+			return err
+		}
+		hdr.c2[i] = newMatrixG2(0, 0)
+		err = hdr.c2[i].unmarshalBinary(c2data)
+		if err != nil {
+			return err
+		}
+	}
+
+	c3Len := int(binary.LittleEndian.Uint16(data))
+	hdr.c3 = make([]*matrixG1, c3Len)
+	hdr.c3neg = make([]*matrixG1, c3Len)
+	data = data[2:]
+
+	for i := 0; i < c3Len; i++ {
+		c3data, data, err = removeLenPrefixed(data)
+		if err != nil {
+			return err
+		}
+		hdr.c3[i] = newMatrixG1(0, 0)
+		err = hdr.c3[i].unmarshalBinary(c3data)
+		if err != nil {
+			return err
+		}
+	}
+
+	for i := 0; i < c3Len; i++ {
+		c3negdata, data, err = removeLenPrefixed(data)
+		if err != nil {
+			return err
+		}
+
+		if len(c3negdata) != 0 {
+			hdr.c3neg[i] = newMatrixG1(0, 0)
+			err = hdr.c3neg[i].unmarshalBinary(c3negdata)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func GenerateParams(rnd io.Reader) (*PublicParams, *SecretParams, error) {
+	A, err := sampleDlin()
+	if err != nil {
+		return nil, nil, err
+	}
+	Bbar, err := randomMatrixZp(4, 4)
```

Comment:
here the `rnd` variable should be propagated to all functions that use randomness.
and the same applies to other functions using a random source.